tests/5.1/target/test_target_has_device_addr.c

```diff
@@ -0,0 +1,54 @@
+//===--- test_target_has_device_addr.c --------------------------------------===//
+//
+// OpenMP API Version 5.1 Nov 2020
+//
+// This test verifies the 'has_device_addr' feature added to the target construct.
+// We tested this by mapping a scalar & array to the device
+// and ensuring that the address does not change when using
+// has_device_addr on another target region.
+//
+////===----------------------------------------------------------------------===//
+#include <assert.h>
+#include <omp.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include "ompvv.h"
+
+#define N 1000
+
+//test by mapping to device use 'target map' construct
+int test_target_has_device_addr() {
+  OMPVV_INFOMSG("test_target_has_device_addr");
+  int errors = 0;
+  int x = 10;
+  int arr[N];
+  for(int i=0;i<N;i++){
+    arr[i] = i;
+  }
+  int *first_scalar_device_addr;
+  int *first_arr_device_addr;
+  #pragma omp target enter data map(to: x, arr)
+  #pragma omp target map(from: first_scalar_device_addr, first_arr_device_addr) map(to: x, arr)
+  {
+    first_scalar_device_addr = &x;
+    first_arr_device_addr = &arr[0];
+  }
+  int *second_scalar_device_addr, *second_arr_device_addr;
+  //check addresses are same on device region
```

Comment:
@nolanbaker31 wrote:
> No LLVM implementation yet, not sure about GCC 12

GCC12 supports `has_device_addr` in C/C++/Fortran.

But testing showed that the testcase fails for second_..._device_addr. The problem is that the code claims that x/arr has the device address – but it actually has the host address.

How about:

```suggestion
  //check addresses are same on device region
  #pragma omp target map(from:second_scalar_device_addr, second_arr_device_addr) has_device_addr(x, arr) 
```